### PR TITLE
LR-warmup-steps sensitivity sweep (4k vs 8k vs 11k steps)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

PR #994 (tanjiro, branch `tanjiro/lr-warmup-decouple-from-vol-schedule`) introduced `--lr-warmup-steps` for step-granular LR warmup that decouples the warmup completion from the EP0→EP1 vol_points curriculum jump. Run `p3v6veyl` in that PR showed an **exceptional result**: `val_abupt=7.44%, val_vol_p=4.83%` at step 18,990 — both EP3 gate conditions already met with ~13,600 steps to spare. This is the strongest 4-epoch screen result we have seen.

The calibrated default used in #994 Arm A was `--lr-warmup-steps 8000` (completing warmup at ~74% of EP0, i.e. step 8,000 of 10,864 steps). **We don't yet know the sensitivity of this hyperparameter.** A warmup that's too short may still cause the EP1 shock; one that's too long may compress useful cosine annealing time. The sweet spot is unknown.

**This PR sweeps `--lr-warmup-steps` across three values** to find the optimal warmup length:
- **Arm A**: `--lr-warmup-steps 4000` (≈37% through EP0 — short warmup, LR arrives at peak well before vol jump)
- **Arm B**: `--lr-warmup-steps 8000` (≈74% through EP0 — the calibrated #994 value, replicate for comparison)
- **Arm C**: `--lr-warmup-steps 11000` (≈101% of EP0, i.e. warmup extends 136 steps into EP1 — long warmup)

Steps/epoch reference (EP0): 400 train cases × 65536 surf pts ÷ (8 GPUs × 4 batch_size) = **10,864 steps** with `vol_pts=16,384` fast schedule.

## Prerequisite

This experiment depends on **PR #994 being merged** first, as it adds the `--lr-warmup-steps` flag to `train.py`. Check that the flag exists before running:

```bash
cd target/ && python train.py --help | grep lr-warmup-steps
```

If the flag is not present, **wait for PR #994 to merge** and pull the latest `tay` branch before starting.

## Implementation

**No code changes required beyond what PR #994 adds.** Once #994 is merged, you simply change the `--lr-warmup-steps` value between arms. All other hyperparameters remain identical to the #994 SOTA run.

## Training Commands

All three arms use the fast vol-curriculum and 4-epoch screen schedule. Run Arms A and B in parallel first; if one clearly dominates by EP2 kill the other.

### Arm A — Short warmup (4000 steps):

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 \
  --lr-warmup-steps 4000 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-vol-pressure-aux-head --volume-loss-weight 1.0 \
  --wandb-group edward-lr-warmup-steps-sweep
```

### Arm B — Calibrated warmup (8000 steps, replicate of #994):

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 \
  --lr-warmup-steps 8000 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-vol-pressure-aux-head --volume-loss-weight 1.0 \
  --wandb-group edward-lr-warmup-steps-sweep
```

### Arm C — Long warmup (11000 steps):

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 \
  --lr-warmup-steps 11000 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-vol-pressure-aux-head --volume-loss-weight 1.0 \
  --wandb-group edward-lr-warmup-steps-sweep
```

Run Arm C after Arms A/B have shown at least EP2 results — use the remaining 3 GPUs if available, or queue it after the first arms complete.

## EP3 Kill Gate

Kill any arm at the end of EP3 if **both** conditions are NOT met:
- `val_abupt ≤ 8.0%`
- `val_vol_p ≤ 5.0%`

The tanjiro #994 run met both at step 18,990 (~EP1.7). If arms here show similar trajectories, let them run through EP4.

## Baseline

Current best metrics (PR #958, run `29nohj67`):
| Metric | Value |
|--------|-------|
| val_abupt | 6.2868% |
| test_abupt | 7.7445% |
| test_vol_p | 12.0063% |

Reference tanjiro #994 4-epoch screen (run `p3v6veyl`):
| Metric | Value |
|--------|-------|
| val_abupt | 7.44% at step 18,990 |
| val_vol_p | 4.83% at step 18,990 |

**Gate to beat for full training**: `val_abupt < 6.2868%`

W&B project: `wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`

## What to Report

For each arm, report:
1. EP3 val_abupt and val_vol_p
2. The step at which both gate conditions were first met (if at all)
3. EP4 final val_abupt and val_vol_p (if EP3 gate passed)
4. W&B run IDs for all arms

## Suggested Follow-ups

- Winning warmup value should be used as the new default for all future long-training runs
- If Arm C (11000) wins: explore even longer warmup e.g. steps=15000 or full 2-epoch warmup
- Combine winning warmup with alphonse's OOD-4 upweighting if both show improvement

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT edward:

## Prerequisite check — PR #994 not yet merged

The PR body says to wait for PR #994 to merge before starting. Status check:

- **#994 state**: OPEN, draft, status:wip (currently running 13-ep confirmation run, EP2 PASS as of latest comment)
- **#994 mergeable**: CLEAN
- **`--lr-warmup-steps` in my branch**: NOT PRESENT (verified `python train.py --help | grep warmup` only shows `--lr-warmup-epochs`)
- **Diff scope of #994 implementation**: small — `train.py` +13 lines (flag plumbing + per-batch scheduler.step), `trainer_runtime.py` +25 lines (step-granular scheduler)

I have two options:

**Option 1 — Wait for #994 to merge into `tay`** (per PR instructions). I exit and the entrypoint re-invokes me when #994 is merged and `tay` advances.

**Option 2 — Rebase #1020 on `tanjiro/lr-warmup-decouple-from-vol-schedule` and start immediately.** The implementation is settled (Arm A EP2 PASS confirms the scheduler logic works). Risk: if #994 amends the implementation before merge, I'd need to rebase and rerun.

Which would you prefer? Default is Option 1 unless you say otherwise.

**Comment by morganmcg1** [comment]
ADVISOR: Good catch on the flag dependency. Go with **Option 1 — wait for PR #994 to merge**. PR #994 has EP2 PASS confirmed and is in clean mergeable state, so the wait should be short. Rebasing on tanjiro's in-flight branch risks divergence if he amends before merge.

Once #994 is merged, rebase your branch on the updated main and your sweep will have the correct `--lr-warmup-steps` flag available. The three-arm sweep (4000/8000/11000) is exactly what we need to calibrate optimal warmup length.

Hold here and ping me when #994 lands.

**Comment by morganmcg1** [comment]
STUDENT: 

## PR #994 was CLOSED (not merged) — sweep prerequisite gone

Status check:
- **PR #994**: state=CLOSED at 2026-05-12T04:19:00Z, mergedAt=null, mergeCommit=null
- **`--lr-warmup-steps` flag**: NOT present on `tay` (verified `python train.py --help | grep warmup` only shows `--lr-warmup-epochs` and `--gradnorm-init-warmup-steps`)
- The flag still lives on the unmerged branch `tanjiro/lr-warmup-decouple-from-vol-schedule` (commit `bc2afdb`)

**Why #994 was closed** (from your closing comment 2026-05-12T04:18:59Z):
> The best result from this PR (val_abupt=6.6565%, test_abupt=8.017%) is **0.22pp behind the SOTA** in val and 0.32pp behind in test_abupt. The primary metric test_vol_p (12.25%) also regressed from the SOTA baseline (~11.67%). [...] The hypothesis that the LR warmup/vol_points collision at EP1→EP2 is the primary bottleneck for SOTA convergence is **not confirmed**. [...] The `--lr-warmup-steps` flag is a useful infrastructure addition (cherry-pick candidate for future experiments that need fine-grained warmup control), but it does not improve on the current SOTA recipe. [...] GPU time is better directed toward encoder-level OOD interventions (PRs #1019, #1017, #1024).

Given this, my sweep (4k/8k/11k step warmup) is now investigating a hyperparameter that was already determined not to improve on epoch-based warmup. The 8k value was the calibrated one and it lost to SOTA; tuning around it on the same 4-ep fast-schedule screen is unlikely to flip that verdict.

**Three options — please advise:**

**Option 1 — Close this PR as superseded by #994's negative finding.** Cleanest path; aligns with your stated GPU-time priorities (OOD interventions in #1019/#1017/#1024). Default unless you say otherwise.

**Option 2 — Cherry-pick `bc2afdb` (the flag implementation) onto this branch and run the 3-arm sweep anyway.** Cost ~3×4ep runs (~12 GPU-hours total on 8 GPUs, 3 sequential or 3 simultaneously if I split GPUs). Value: definitive answer on warmup-length sensitivity at the fast-schedule 4-ep regime. Risk: even a best arm here is unlikely to recover the 0.22pp test_abupt SOTA gap given the slow-schedule 13-ep result was already negative.

**Option 3 — Pivot the PR.** If there's a related question you'd like me to investigate using the cherry-picked flag (e.g. step-warmup combined with one of the OOD interventions, or a different schedule pairing), let me know what variant would actually be informative.

Holding here. Will close if I don't hear back and you re-assign me elsewhere.

**Comment by morganmcg1** [comment]
ADVISOR: PR #994 has been **closed as a negative result** — the `--lr-warmup-steps` flag was NOT merged into `tay`. Your sweep as designed is no longer executable without implementing the flag from scratch.

**Summary of PR #994 outcome:** The best result from the 13-epoch confirmation run was val_abupt=6.6565%, test_vol_p=12.25% — both 0.22pp worse than SOTA on val_abupt and regressing on the primary metric. The step-based warmup hypothesis was not confirmed. There is no value in running the 4k/8k/11k sweep.

**Recommendation: Close this PR and pivot to a new hypothesis.** I will assign you a fresh experiment shortly. Reply to confirm you are OK closing this PR — or if you'd prefer to pivot the PR body yourself to a new hypothesis, propose one and I'll review it. Either way, do not start any new training until I send the next assignment.

**Comment by morganmcg1** [comment]
STUDENT: Confirmed — closing this PR. The 4k/8k/11k sweep is no longer executable (prerequisite PR #994 closed without merging the `--lr-warmup-steps` flag), and the underlying step-based warmup hypothesis was already disconfirmed by #994's 13-ep confirmation run (val_abupt=6.6565% vs SOTA 6.2868%; test_vol_p=12.25% regression).

Holding for the next assignment per your instruction — will not start any new training. Standing by.

**Comment by morganmcg1** [comment]
Closing as superseded by PR #994's negative finding. The `--lr-warmup-steps` flag was not merged to `tay`, and the calibrated 8k value already lost to SOTA in #994's 13-ep confirmation. GPU time better directed toward OOD interventions per advisor.
